--- 4.6.1/src/corelib/io/qprocess_win.cpp       2010-07-19 13:32:56.009500000 -0700
+++ 4.6.1_64/src/corelib/io/qprocess_win.cpp    2012-01-26 16:59:07.376668800 -0800
@@ -383,7 +383,7 @@
                             (wchar_t*)args.utf16(),
                             0, 0, false, 0, 0, 0, 0, pid);
 #else
-    DWORD dwCreationFlags = CREATE_NO_WINDOW;
+    DWORD dwCreationFlags = 0;//CREATE_NO_WINDOW;
     dwCreationFlags |= CREATE_UNICODE_ENVIRONMENT;
     STARTUPINFOW startupInfo = { sizeof( STARTUPINFO ), 0, 0, 0,
                                  (ulong)CW_USEDEFAULT, (ulong)CW_USEDEFAULT,
